compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -734,13 +733,36 @@ abstract class GenJSCode extends plugins.PluginComponent
         if (scalaPrimitives.isPrimitive(sym)) {
           None
         } else if (sym.isDeferred || sym.owner.isInterface) {
+          val body = if (sym.hasAnnotation(JavaDefaultMethodAnnotation)) {
+            /* For an interface method with @JavaDefaultMethod, make it a
+             * default method calling the impl class method.
+             */
+            val implClassSym = sym.owner.implClass
+            val implMethodSym = implClassSym.info.member(sym.name).suchThat { s =>
+              s.isMethod &&
+              s.tpe.params.size == sym.tpe.params.size + 1 &&
+              s.tpe.params.head.tpe =:= sym.owner.toTypeConstructor &&
+              s.tpe.params.tail.zip(sym.tpe.params).forall {
+                case (sParam, symParam) =>
+                  sParam.tpe =:= symParam.tpe
+              }
+            }
+            genTraitImplApply(implMethodSym,
+                js.This()(currentClassType) :: jsParams.map(_.ref))
+          } else {
+            js.EmptyTree
+          }
           Some(js.MethodDef(static = false, methodName,
-              jsParams, currentClassType, js.EmptyTree)(
+              jsParams, toIRType(sym.tpe.resultType), body)(
```

Comment:
I do not understand this... Was the return type set to the class type??? That's just plain wrong?!?